File: kotlin/workers/src/io/bazel/ruleskotlin/workers/compilers/jvm/utils/JdepsParser.java
/*
 * Copyright 2018 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.bazel.ruleskotlin.workers.compilers.jvm.utils;

import com.google.devtools.build.lib.view.proto.Deps;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Stream;

public class JdepsParser {
    private final String filename;
    private final String packageSuffix;
    private final Predicate<String> isImplicit;

    private final Map<String, Deps.Dependency.Builder> depMap = new HashMap<>();
    private final Set<String> packages = new HashSet<>();

    private JdepsParser(String filename, Predicate<String> isImplicit) {
        this.filename = filename;
        this.packageSuffix = " (" + filename + ")";
        this.isImplicit = isImplicit;
    }

    private void consumeJarLine(String classJarPath, Deps.Dependency.Kind kind) {
        Path path = Paths.get(classJarPath);

        // ignore absolute files, -- jdk jar paths etc.
        // only process jar files
        if (!(path.isAbsolute() || !classJarPath.endsWith(".jar"))) {
            Deps.Dependency.Builder entry = depMap.computeIfAbsent(classJarPath, (key) -> {
                Deps.Dependency.Builder depBuilder = Deps.Dependency.newBuilder();
                depBuilder.setPath(classJarPath);
                depBuilder.setKind(kind);

                if (isImplicit.test(classJarPath)) {
                    depBuilder.setKind(Deps.Dependency.Kind.IMPLICIT);
                }
                return depBuilder;
            });

            // don't flip an implicit dep.
            if (entry.getKind() != Deps.Dependency.Kind.IMPLICIT) {
                entry.setKind(kind);
            }
        }
    }

    private enum Mode {
        COLLECT_DEPS,
        DETERMINE_JDK,
        COLLECT_PACKAGES_JDK8,
        COLLECT_PACKAGES_JDK9
    }

    private Mode mode = Mode.COLLECT_DEPS;

    // maybe simplify this by tokenizing on whitespace and arrows.
    private void processLine(String line) {
        String trimmedLine = line.trim();
        switch (mode) {
            case COLLECT_DEPS:
                if (!line.startsWith(" ")) {
                    String[] parts = line.split(" -> ");
                    if (parts.length == 2) {
                        if (!parts[0].equals(filename)) {
                            throw new RuntimeException("should only get dependencies for dep: " + filename);
                        }
                        consumeJarLine(parts[1], Deps.Dependency.Kind.EXPLICIT);
                    }
                } else {
                    mode = Mode.DETERMINE_JDK;
                    processLine(line);
                }
                break;
            case DETERMINE_JDK:
                mode = Mode.COLLECT_PACKAGES_JDK8;
                if (!line.endsWith(packageSuffix)) {
                    mode = Mode.COLLECT_PACKAGES_JDK9;
                }
                processLine(line);
                break;
            case COLLECT_PACKAGES_JDK8:
                if (trimmedLine.endsWith(packageSuffix)) {
                    packages.add(trimmedLine.substring(0, trimmedLine.length() - packageSuffix.length()));
                } else if (trimmedLine.startsWith("-> ")) {
                    // ignore package detail lines, in the jdk8 format these start with arrows.
                } else throw new RuntimeException("unexpected line while collecting packages: " + line);
                break;
            case COLLECT_PACKAGES_JDK9:
                String[] pkg = trimmedLine.split("\\s+");
                packages.add(pkg[0]);
                break;
        }
    }


    public static Predicate<String> pathSuffixMatchingPredicate(Path directory, String... jars) {
        String[] suffixes = Stream.of(jars).map(lib -> directory.resolve(lib).toString()).toArray(String[]::new);
        return (jar) -> {
            for (String implicitJarsEnding : suffixes) {
                if (jar.endsWith(implicitJarsEnding)) {
                    return true;
                }
            }
            return false;
        };
    }

    public static Deps.Dependencies parse(String label, String classJar, String classPath, Stream<String> jdepLines, Predicate<String> isImplicit) {
        String filename = Paths.get(classJar).getFileName().toString();
        JdepsParser jdepsParser = new JdepsParser(filename, isImplicit);
        Stream.of(classPath.split(":")).forEach(x -> jdepsParser.consumeJarLine(x, Deps.Dependency.Kind.UNUSED));
        jdepLines.forEach(jdepsParser::processLine);

        Deps.Dependencies.Builder rootBuilder = Deps.Dependencies.newBuilder();
        rootBuilder.setSuccess(false);
        rootBuilder.setRuleLabel(label);

        rootBuilder.addAllContainedPackage(jdepsParser.packages);
        jdepsParser.depMap.values().forEach(b -> rootBuilder.addDependency(b.build()));

        rootBuilder.setSuccess(true);
        return rootBuilder.build();
    }
}

File: kotlin/builder/src/io/bazel/kotlin/builder/mode/jvm/utils/JdepsParser.kt
/*
 * Copyright 2018 The Bazel Authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.bazel.kotlin.builder.mode.jvm.utils

import com.google.devtools.build.lib.view.proto.Deps
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.function.Predicate
import java.util.stream.Stream

class JdepsParser private constructor(private val filename: String, private val isImplicit: Predicate<String>) {
    private val packageSuffix: String = " ($filename)"

    private val depMap = HashMap<String, Deps.Dependency.Builder>()
    private val packages = HashSet<String>()

    private var mode = Mode.COLLECT_DEPS

    private fun addJar(classJarPath: String, kind: Deps.Dependency.Kind) {
        val path = Paths.get(classJarPath)

        // ignore absolute files, -- jdk jar paths etc.
        // only process jar files
        if (!(path.isAbsolute || !classJarPath.endsWith(".jar"))) {
            val entry = depMap.computeIfAbsent(classJarPath) {
                val depBuilder = Deps.Dependency.newBuilder()
                depBuilder.path = classJarPath
                depBuilder.kind = kind

                if (isImplicit.test(classJarPath)) {
                    depBuilder.kind = Deps.Dependency.Kind.IMPLICIT
                }
                depBuilder
            }

            // don't flip an implicit dep.
            if (entry.kind != Deps.Dependency.Kind.IMPLICIT) {
                entry.kind = kind
            }
        }
    }

    private enum class Mode {
        COLLECT_DEPS,
        DETERMINE_JDK,
        COLLECT_PACKAGES_JDK8,
        COLLECT_PACKAGES_JDK9
    }

    // maybe simplify this by tokenizing on whitespace and arrows.
    private fun processLine(line: String) {
        val trimmedLine = line.trim { it <= ' ' }
        when (mode) {
            Mode.COLLECT_DEPS -> if (!line.startsWith(" ")) {
                val parts = line.split(" -> ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                if (parts.size == 2) {
                    if (parts[0] != filename) {
                        throw RuntimeException("should only get dependencies for dep: $filename")
                    }
                    addJar(parts[1], Deps.Dependency.Kind.EXPLICIT)
                }
            } else {
                mode = Mode.DETERMINE_JDK
                processLine(line)
            }
            Mode.DETERMINE_JDK -> {
                mode = Mode.COLLECT_PACKAGES_JDK8
                if (!line.endsWith(packageSuffix)) {
                    mode = Mode.COLLECT_PACKAGES_JDK9
                }
                processLine(line)
            }
            Mode.COLLECT_PACKAGES_JDK8 -> when {
                trimmedLine.endsWith(packageSuffix) -> packages.add(trimmedLine.substring(0, trimmedLine.length - packageSuffix.length))
                trimmedLine.startsWith("-> ") -> {
                    // ignore package detail lines, in the jdk8 format these start with arrows.
                }
                else -> throw RuntimeException("unexpected line while collecting packages: $line")
            }
            Mode.COLLECT_PACKAGES_JDK9 -> {
                val pkg = trimmedLine.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                packages.add(pkg[0])
            }
        }
    }

    companion object {
        fun pathSuffixMatchingPredicate(directory: Path, vararg jars: String): Predicate<String> {
            val suffixes = jars.map { directory.resolve(it).toString() }
            return Predicate { jar ->
                for (implicitJarsEnding in suffixes) {
                    if (jar.endsWith(implicitJarsEnding)) {
                        return@Predicate true
                    }
                }
                false
            }
        }

        fun parse(label: String, classJar: String, classPath: String, jdepLines: Stream<String>, isImplicit: Predicate<String>): Deps.Dependencies {
            val filename = Paths.get(classJar).fileName.toString()
            val jdepsParser = JdepsParser(filename, isImplicit)

            Stream.of(*classPath.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()).forEach {
                jdepsParser.addJar(it, Deps.Dependency.Kind.UNUSED)
            }

            jdepLines.forEach { jdepsParser.processLine(it) }

            val rootBuilder = Deps.Dependencies.newBuilder()
            rootBuilder.success = false
            rootBuilder.ruleLabel = label

            rootBuilder.addAllContainedPackage(jdepsParser.packages)
            jdepsParser.depMap.values.forEach { b -> rootBuilder.addDependency(b.build()) }

            rootBuilder.success = true
            return rootBuilder.build()
        }
    }
}